dense_vector/track.py

```diff
@@ -99,45 +176,37 @@ def params(self):
 # the accuracy of the knn query.
 class KnnRecallRunner:
     async def __call__(self, es, params):
+        k = params["size"]
+        num_candidates = params["num_candidates"]
+        index = params["index"]
+        request_cache = params["cache"]
+        queries_file = params["queries_file"]
+        vector_field = params["vector_field"]
+        target_k = max(params["target_k"], k)
         recall_total = 0
         exact_total = 0
-        min_recall = params["size"]
+        min_recall = k
 
-        for query in params["queries"]:
+        knn_vector_store: KnnVectorStore = KnnVectorStore.get_instance(queries_file, vector_field)
```

Comment:
We kind of discussed this already via Slack, but I think you'd be better off moving this `KnnVectorStore` initialisation out of the `KnnRecallRunner` runner and into to the `KnnRecallParamSource` param source constructor. That way we cache the file upfront before we even start issuing requests. 

I know you mentioned that the latency of these requests isn't a big concern, so I'll leave it up to you to decide given this is your workload, but in general we try and avoid adding anything blocking/slow to a runner besides the actual operation we're trying to measure.

From Slack:
>

Reply:
Tbh I kept it to `KnnRecallRunner` because in addition to being interested only in recall, IIUC, based on logs, initializing `KnnRecallParamSource` happens a number of times throughout the track's execution (e.g. for each recall operation) so I tried avoiding the additional I/O. We do however also get the plus of caching the knn vector store, as this happens only on two separate actors.E.g. logs from adding the `KnnVectorStore` to the constructor of `KnnRecallParamSource` **with** the lru cache:
```
❯ tail -f /Users/panagiotis.bailis/.rally/logs/rally.log | grep "Initializing KnnVectorStore"                                                                                          ─╯
2023-11-24 15:56:52,129 ActorAddr-(T|:59518)/PID:70822 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 15:57:00,732 ActorAddr-(T|:59571)/PID:70835 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
```
and **without** the lru cache:
```
❯ tail -f /Users/panagiotis.bailis/.rally/logs/rally.log  | grep "Initializing KnnVectorStore"                                                                                         ─╯
2023-11-24 16:39:10,13 ActorAddr-(T|:62100)/PID:75072 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:10,52 ActorAddr-(T|:62100)/PID:75072 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:10,89 ActorAddr-(T|:62100)/PID:75072 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:10,126 ActorAddr-(T|:62100)/PID:75072 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:10,163 ActorAddr-(T|:62100)/PID:75072 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:18,794 ActorAddr-(T|:62150)/PID:75085 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:39:41,863 ActorAddr-(T|:62150)/PID:75085 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:40:06,430 ActorAddr-(T|:62150)/PID:75085 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:40:27,997 ActorAddr-(T|:62150)/PID:75085 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
2023-11-24 16:40:51,577 ActorAddr-(T|:62150)/PID:75085 dense_vector.track INFO Initializing KnnVectorStore for queries file: '/Users/panagiotis.bailis/workspace/github/pmpailis/rally-tracks/dense_vector/queries.json' and vector field: 'vector'
```

Would it make sense to add the params on the `KnnRecallParamSource` constructor and then read the file "lazily" when actually calling the `params` and passing it to the runner? This would lead to the file been read only once instead for each operation. I've made some changes that address this so whenever you have some time could you please give it another 👀 ? 🙏

Reply:
Hmm, we shouldn't be constructing a `KnnRecallParamSource` more than once per worker (actor):
https://github.com/elastic/rally/blob/c1f04a368f5efaa28128dc5341b7082a62a13a34/esrally/driver/driver.py#L1792-L1812

In any case this is your workload/track, so if this implementation makes sense then I have no problems with it, just wanted to point out the critical path bit.

Reply:
> Hmm, we shouldn't be constructing a KnnRecallParamSource more than once per worker (actor):

IIUC doesn't this codepath execute independently for each operation that the actor is assigned to, as we initialize a new `AsyncIoAdapter` for each task allocation - and the `params_per_task` are independent for each run (which makes sense since we have different params per operation)?  🤔   Maybe I'm missing something but - do we already account somehow for having commonly shared resources for similar operations? 

In addition to that, we also seem to generate a `KnnRecallParamSource` when preparing the track and check for [used corpora](https://github.com/elastic/rally/blob/c1f04a368f5efaa28128dc5341b7082a62a13a34/esrally/track/loader.py#L442) which also introduces (for the specific case) unnecessary I/O load.  

(_Note:_  I think that the same holds for `KnnParamSource` as well)

Reply:
> IIUC doesn't this codepath execute independently for each operation that the actor is assigned to, as we initialize a new AsyncIoAdapter for each task allocation 

Yes, sorry, I glossed over that part, I should have said "once per worker, per task".

> and the params_per_task are independent for each run (which makes sense since we have different params per operation)? 🤔 Maybe I'm missing something but - do we already account somehow for having commonly shared resources for similar operations?

[The `AsyncExecutor` is the part the actually executes the runner](https://github.com/elastic/rally/blob/c1f04a368f5efaa28128dc5341b7082a62a13a34/esrally/driver/driver.py#L1905-L1928), and that gets its [per-request parameters from the respective `ScheduleHandle`](https://github.com/elastic/rally/blob/c1f04a368f5efaa28128dc5341b7082a62a13a34/esrally/driver/driver.py#L2369-L2452). 

I took a deeper look at the code but I lack almost all of the required domain knowledge here to make any good advice. I'll leave it up to you how you want to handle it, i.e. you don't need my approval to merge 😆 as these are just general comments.